packages/ref-core/src/ref_core/constraints.py

```diff
@@ -0,0 +1,148 @@
+        """
+        Validate if the constraint is satisfied by the dataset.
+
+        This is executed after the apply method to determine if the constraint is satisfied.
+        If the constraint is not satisfied, the group will not be executed.
+
+        Parameters
+        ----------
+        group
+            A group of datasets that is being validated.
+
+        Returns
+        -------
+        :
+            Whether the constraint is satisfied
+        """
+        ...
+
+
+@runtime_checkable
+class GroupOperation(Protocol):
+    """
+    An operation to perform on a group of datasets resulting in a new group of datasets.
+    """
+
+    def apply(self, group: pd.DataFrame, data_catalog: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform an operation on the group of datasets.
+
+        If this operation modifies the group, the modified group should be returned.
+        If this operation fails, a ConstraintNotSatisfied exception should be raised.
+
+        Parameters
+        ----------
+        group
+            A group of datasets that is being validated.
+        data_catalog
+            The data catalog of datasets
+
+        Raises
+        ------
+        ConstraintNotSatisfied
+            The operation was not successful
+
+        Returns
+        -------
+        :
+            The updated group of datasets
+        """
+        ...
+
+
+Constraint = GroupOperation | GroupValidator
+"""
+A constraint that must be satisfied when executing a given metric run.
+
+This can include operations that are applied to a group of datasets which may modify the group,
+but may also include validators that check if the group satisfies a certain condition.
+
+All constraints must be satisfied for a given group to be run.
+If any constraint is not satisfied, the group will not be executed.
+"""
+
+
+def apply_constraint(
+    dataframe: pd.DataFrame, constraint: Constraint, data_catalog: pd.DataFrame
+) -> pd.DataFrame | None:
+    """
+    Apply a constraint to a group of datasets
+
+    Parameters
+    ----------
+    constraint
+        The constraint to apply
+    data_catalog
+        The data catalog of datasets
+
+    Returns
+    -------
+    :
+        The updated group of datasets or None if the constraint was not satisfied
+    """
+    try:
+        updated_group = (
+            constraint.apply(dataframe, data_catalog) if isinstance(constraint, GroupOperation) else dataframe
+        )
+
+        valid = constraint.validate(updated_group) if isinstance(constraint, GroupValidator) else True
+        if not valid:
+            logger.debug(f"Constraint {constraint} not satisfied for {dataframe}")
+            raise ConstraintNotSatisfied(f"Constraint {constraint} not satisfied for {dataframe}")
+    except ConstraintNotSatisfied:
+        logger.debug(f"Constraint {constraint} not satisfied for {dataframe}")
+        return None
+
+    return updated_group
+
+
+@frozen
+class RequiredFacets:
```

Comment:
```suggestion
class RequireFacets:
```

For `SelectParentExperiment` you use imperative present, I think that would be good to use everywhere in the Constraint names.